aim/ext/transport/server.py

```diff
@@ -42,46 +45,68 @@ class RemoteTrackingServicer(remote_tracking_pb2_grpc.RemoteTrackingServiceServi
     count = 0
 
     def get_resource(self, request: rpc_messages.ResourceRequest, context):
-        resource_cls = self.registry._registry[request.resource_type]
-        if len(request.args) > 0:
-            res = resource_cls(*request.args)
-        else:
-            res = resource_cls()
-        handler = _get_handler()
-        self.resource_pool[handler] = (request.client_uri, res)
-
-        return rpc_messages.ResourceResponse(handler=handler)
+        resource_handler = _get_handler()
+        try:
+            resource_cls = self.registry[request.resource_type]
+            if len(request.args) > 0:
+                res = resource_cls(*request.args)
+            else:
+                res = resource_cls()
+
+            self.resource_pool[resource_handler] = (request.client_uri, res)
+            return rpc_messages.ResourceResponse(handler=resource_handler,
+                                                 status=rpc_messages.ResourceResponse.Status.OK)
+        except Exception as e:
+            try:
+                # clean up resource_pool
+                # because no one will call release_resource in case of Exception
+                del self.resource_pool[resource_handler]
+            except KeyError:
+                pass
+
+            return rpc_messages.ResourceResponse(
+                handler=None,
+                status=rpc_messages.ResourceResponse.Status.ERROR,
+                exception=build_exception(e),
+            )
 
     def release_resource(self, request: rpc_messages.ReleaseResourceRequest, context):
-        resource_handler = request.handler
-        client_uri = request.client_uri
-        if not self._verify_resource_handler(resource_handler, client_uri):
-            raise UnauthorizedRequestError()
-
-        del self.resource_pool[resource_handler]
-        return rpc_messages.ReleaseResourceResponse(status=0)
+        try:
+            resource_handler = request.handler
+            client_uri = request.client_uri
+            if not self._verify_resource_handler(resource_handler, client_uri):
+                raise UnauthorizedRequestError()
+
+            del self.resource_pool[resource_handler]
+            return rpc_messages.ReleaseResourceResponse(status=rpc_messages.ReleaseResourceResponse.Status.OK)
+        except Exception as e:
+            # Should we clean ed resource_pool in case of exceptions? what if UnauthorizedRequest?
```

Comment:
what if instead of `UnauthorizedRequest` we do not raise exception than catch it and handle but just return `ReleaseResourceResponse` with ERROR status and appropriate exception? so basically in LINE 78:
```
return rpc_messages.ReleaseResourceResponse(
    status=rpc_messages.ReleaseResourceResponse.Status.ERROR,
    exception=build_exception(UnauthorizedRequestError()),
)
```

Reply:
Thanks for notes. I guess `raise` more readable than `return` in this particular case. Plus with `Single Return` we can easily add some clean-up codes or easily change Exception safety of the methods.